### ISSUE TITLE
[FEATURE] Typeless event for AtomBaseVariable

### ISSUE THREAD

**Issue by xvnnnk** [AdamRamberg/unity-atoms#136]
**Is your feature request related to a problem? Please describe.**
We can have AtomBaseVariable as a serialized field, but there is no way to access its `Changed` event. 
Since all AtomBaseVariables have Values and that value can change, it is possible to have a typeless Changed event.

**Describe the solution you'd like**
To have `TypelessChanged` event in `AtomBaseVariable`.

**Describe alternatives you've considered**
The only alternative is to manually track changes of AtomBaseVariable value every frame in script that uses it.

**Comment by soraphis** [comment]
bind

**Comment by soraphis** [comment]
> The only alternative is to manually track changes of AtomBaseVariable value every frame in script that uses it.

that is not true. not at all. see this example:

```
public class HealthBar : MonoBehaviour
{
    [SerializeField]
    private IntVariable MaxHealth;

    void Start()
    {
        MaxHealth.Changed.Register((oldValue) => this.DoSomething());
    }
    
    // todo: handle unregister!
   
    private void DoSomething()
    {
    }
}
```

a drawback of this method is, you need to hook it up via code. also you need to find a way of handling the unregister, better to save the lambda method in a variable. 

or just use a method with a parameter, but ignore the parameter.

you **do not** need to check every frame. 


other alternative: register a void event to raise when your TYPED-Event is raised

**Comment by soraphis** [comment]
just checked, and the Register call allows `Action` as delegate.

so calling register via code you can use any `void method()` you want. 

in the example above this would mean:

        MaxHealth.Changed.Register(DoSomething);

**Comment by mnarimani** [comment]
@soraphis 
This is the correct example:
```
public class Script : MonoBehaviour 
{
    public AtomBaseVariable Value;

    private void Start()
    {
        // How to register to AtomBaseVariable?
    }
}
```

**Comment by soraphis** [comment]
> We can have AtomBaseVariable as a serialized field, but there is no way to access its Changed event.

AtomBaseVariable **has no** changed event. so there obviously is no way to access it.

**Comment by mnarimani** [comment]
> > We can have AtomBaseVariable as a serialized field, but there is no way to access its Changed event.
> 
> AtomBaseVariable **has no** changed event. so there obviously is no way to access it.

Hence this feature request.
There's no problem with adding a typeless event to AtomBaseVariable. As a matter of fact, Scriptable Object Architecture library (you can find it on github) already does this and it's very useful when you want to have a script that just shows content of a AtomBaseVariable in UI.

**Comment by AdamRamberg** [comment]
I don't think we want to add a typeless event to `AtomBaseVariable`, since it is used as a base class for constants as well. Why not just cast `AtomBaseVariable` to whatever type you are expecting and then access the `Changed` event? And why do you need to specify the variable as a `AtomBaseVariable` instead of its specific type, eg. `IntVariable`?

**Comment by mnarimani** [comment]
> I don't think we want to add a typeless event to `AtomBaseVariable`, since it is used as a base class for constants as well. Why not just cast `AtomBaseVariable` to whatever type you are expecting and then access the `Changed` event? And why do you need to specify the variable as a `AtomBaseVariable` instead of its specific type, eg. `IntVariable`?

Because I'm trying to create a general class that shows the content of `AtomBaseVariable` in UI.
If I use a specific type I have to create a class for each type. Which is undesirable.

**Comment by soraphis** [comment]
As @AdamRamberg wrote: cast AtomBaseVariable.

public class Script : MonoBehaviour 
{
    public AtomBaseVariable Value;

    private void Start()
    {
        if(Value is IGetEvent changeable) changeable.GetEvent<AtomEventBase>().Register(() => Debug.Log("something changed"));
    }
}

**Comment by AdamRamberg** [comment]
@soraphis suggestion combined with #140 should solve your case (#140 will be included in v4.2 of Unity Atoms). Closing this issue. If this solution is not sufficient please reopen this issue.
